scripts/searchindex.ts

```diff
@@ -0,0 +1,211 @@
+const { Client } = require('@elastic/elasticsearch');
+const dotenv = require('dotenv');
+
+dotenv.config({ path: '.env.elastic' }); // ignore errors
+
+const mappings = {
+  dynamic: 'false',
+  properties: {
+    body_content: {
+      type: 'text',
+      term_vector: 'with_positions_offsets',
+      fields: {
+        raw: {
+          type: 'keyword',
+        },
+      },
+    },
+    id: {
+      type: 'keyword',
+    },
+    date: {
+      type: 'date',
+    },
+    tags: {
+      type: 'text',
+      fields: {
+        raw: {
+          type: 'keyword',
+        },
+      },
+    },
+    title: {
+      type: 'text',
+      fields: {
+        raw: {
+          type: 'keyword',
+        },
+      },
+    },
+    summary: {
+      type: 'text',
+      fields: {
+        raw: {
+          type: 'keyword',
+        },
+      },
+    },
+    url: {
+      type: 'text',
+      fields: {
+        raw: {
+          type: 'keyword',
+        },
+      },
+    },
+    url_host: {
+      type: 'text',
+      fields: {
+        raw: {
+          type: 'keyword',
+        },
+      },
+    },
+    url_path: {
+      type: 'text',
+      fields: {
+        raw: {
+          type: 'keyword',
+        },
+      },
+    },
+    url_path_dir1: {
+      type: 'text',
+      fields: {
+        raw: {
+          type: 'keyword',
+        },
+      },
+    },
+    url_path_dir2: {
+      type: 'text',
+      fields: {
+        raw: {
+          type: 'keyword',
+        },
+      },
+    },
+    url_path_dir3: {
+      type: 'text',
+      fields: {
+        raw: {
+          type: 'keyword',
+        },
+      },
+    },
+    url_port: {
+      type: 'text',
+      fields: {
+        raw: {
+          type: 'keyword',
+        },
+      },
+    },
+  },
+};
+
+const client = new Client({
+  node: {
+    url: new URL(process.env.ELASTIC_URL),
+  },
+  auth: {
+    username: process.env.ELASTIC_USERNAME,
+    password: process.env.ELASTIC_PASSWORD,
+  },
+  maxRetries: 5,
+  requestTimeout: 60000,
+  sniffOnStart: true,
+});
+
+function getCurrentDate(): string {
+  const now = new Date();
+  const year = now.getFullYear().toString().substr(-2);
+  const month = (now.getMonth() + 1).toString().padStart(2, '0');
+  const day = now.getDate().toString().padStart(2, '0');
+  const hours = now.getHours().toString().padStart(2, '0');
+  const minutes = now.getMinutes().toString().padStart(2, '0');
+  return `${month}${year}${day}${hours}${minutes}`;
+}
+
+(async () => {
+  const { allBlogs } = await import('../.contentlayer/generated/index.mjs');
+  const { execa } = await import('execa');
+
+  const index = `site-search-dev-blog-${getCurrentDate()}`;
+
+  try {
+    await client.indices.create({
+      index,
```

Comment:
This could be in a follow-up since using an alias makes it easy for us to update mappings. But we will want to use an analyzer to provide partial matches.

We can configure the analyzer on the index when we create it, under settings.

```
const settings = {
  index: {
    number_of_shards: 1,
    number_of_replicas: 1,
  },
  analysis: {
    analyzer: {
      my_analyzer: {
        tokenizer: 'my_tokenizer',
      },
    },
    tokenizer: {
      my_tokenizer: {
        type: 'ngram',
        min_gram: 3,
        max_gram: 3,
        token_chars: ['letter', 'digit'],
      },
    },
  },
};

await client.indices.create({
  index,
  mappings,
  settings,
});
```

Then, update the `body_content` in the mapping to make use of it:

```
body_content: {
  type: 'text',
  term_vector: 'with_positions_offsets',
  analyzer: 'partial_match_analyzer',
  fields: {
    raw: {
      type: 'keyword',
    },
  },
},
```

And we should probably update the rest of the fields we are searching across to do the same:

```
body_content: {
  type: 'text',
  term_vector: 'with_positions_offsets',
  analyzer: 'partial_match_analyzer',
  fields: {
    raw: {
      type: 'keyword',
    },
  },
},
tags: {
  type: 'text',
  term_vector: 'with_positions_offsets',
  analyzer: 'partial_match_analyzer',
  fields: {
    raw: {
      type: 'keyword',
    },
  },
},
title: {
  type: 'text',
  term_vector: 'with_positions_offsets',
  analyzer: 'partial_match_analyzer',
  fields: {
    raw: {
      type: 'keyword',
    },
  },
},
summary: {
  type: 'text',
  term_vector: 'with_positions_offsets',
  analyzer: 'partial_match_analyzer',
  fields: {
    raw: {
      type: 'keyword',
    },
  },
},
```